fc/parsing/actions.py

```diff
@@ -780,6 +864,99 @@ def sympy_equations(self):
                 eqs.append(eq.to_sympy(self._symbol_generator, self._number_generator))
         return self._sympy_equations
 
+    #######################################
+    # Helper methods for model manipulation
+
+    def _convert_time_if_needed(self):
+        """Units-convert the time variable if not in the protocol's units."""
+        time_units = self.units.get_unit(self.time_units)
+        time_var = self.model.get_free_variable_symbol()
+        time_var = self.model.convert_variable(time_var, time_units, DataDirectionFlow.INPUT)
+
+    def _add_input_variables(self):
+        """Ensure input variables exist in the desired units.
+
+        TODO: At present this just checks they exist; no units conversion or initial_value handling.
+        TODO: If an input doesn't exist but has units & initial_value here, or is defined via
+        an equation, then add a new variable to the model.
+        """
+        for var in self.inputs:
+            assert var.units is None
+            assert var.initial_value is None
+            self.model.get_symbol_by_ontology_term(var.rdf_term)
+
+    def _add_or_replace_equations(self):
+        """Process define statements and modify the model's equations accordingly."""
+        for eq in self.sympy_equations:
+            lhs = eq.lhs
+            if lhs.is_Derivative:
+                var = lhs.args[0]
+                assert var.initial_value is not None  # TODO: Maybe throw ProtocolError instead?
+            else:
+                var = lhs
+                var.initial_value = None  # In case it was a state variable previously
+            # Figure out if this is a replace or add
+            defn = self.model.get_definition(var)
+            if defn is not None:
+                self.model.remove_equation(defn)
+            self.model.add_equation(eq)
+        # TODO: Check units of newly added equations; apply conversions where needed?
+
+    def _annotate_state_variables(self):
+        """Annotate all state variables with the 'magic' oxmeta:state_variable term."""
+        state_annotation = create_rdf_node((OXMETA_NS, 'state_variable'))
+        self.vector_orderings[state_annotation] = {}
+        for i, state_var in enumerate(self.model.get_state_symbols()):
+            if not state_var.cmeta_id:
+                state_var.cmeta_id = self.model.get_unique_cmeta_id(state_var.name.replace('$', '__'))
```

Comment:
Should we maybe get the "$" sign from cellmlmanip directly?

Reply:
https://github.com/ModellingWebLab/cellmlmanip/issues/228 suggests moving this into a cellmlmanip method anyway.